### PR TITLE
[wallet][rpc]: sendany and tx splitting

### PR DESCRIPTION
**Pull request by nkostoulas**

### CODE REVIEW

**Comment by AstryX** [comment]
> what is the purpose of `std::vector<std::vector<CReserveKey>> vChangeKeys` ?
> 
> odd behaviour:
> 
> ```
>     // Create and send the transaction
>     std::vector<std::vector<CReserveKey>> vChangeKeys;
>     std::vector<CReserveKey> vChangeKey;
>     vChangeKey.reserve(vecSend.size() + 1);
>     for (unsigned int j = 0; j < vecSend.size() + 1; ++j) {
>         vChangeKey.emplace_back(pwalletMain);
>     }
>     vChangeKeys.push_back(vChangeKey);
> 
>     std::vector<CWalletTx> vecRes = pwalletMain->CreateTransaction(vecSend, wtxNew, vChangeKeys, nFeeRequired, nChangePosRet,
>         strError, coinControl, true, NULL, true, NULL, NULL, NULL, CAsset(), fIgnoreBlindFail, fSplitTransactions, std::vector<COutput>(), true);
> 
>     if (!(vecRes.size() > 0)) {
>         if (nValue + nFeeRequired > totalBalance)
>             strError = strprintf("Error: This transaction requires a transaction fee of at least %s", FormatMoney(nFeeRequired));
>         throw JSONRPCError(RPC_WALLET_ERROR, strError);
>     }
>     wtxNew = vecRes[0];
>     for (unsigned int i = 0; i < vecRes.size(); ++i) {
>         CValidationState state;
>         if (!pwalletMain->CommitTransaction(vecRes[i], vChangeKeys[i], g_connman.get(), state)) {
>             strError = strprintf("Error: The transaction was rejected! Reason given: %s %s", state.GetRejectReason(), state.GetDebugMessage());
>             throw JSONRPCError(RPC_WALLET_ERROR, strError);
>         }
>     }
> ```

It is to satisfy the CommitTransaction. The set of keys may be different in size for each CWalletTx if it is split so we can't necessarily reuse the same change at the moment. Additionally, vChangeKeys get unreserved within the code that handles change and it may cause issues if we reused a single vector for everything.

**Comment by nkostoulas** [comment]
Ok @AstryX makes sense. One more question;

Why do you exchange balance between last asset and fee asset:
```
            CRecipient lastRecipient = vecSend.back();
            if (lastRecipient.asset != newFeeAsset) {
                CAmount lastBalance = balanceMap[lastRecipient.asset];
                CAmount usedLastBalance = mapValue[lastRecipient.asset];
                CAmount unusedLastBalance = lastBalance - usedLastBalance;

                mapValue[lastRecipient.asset] = lastBalance;
                mapValue[newFeeAsset] -= unusedLastBalance;

                vecSend[newFeeRecipient].nAmount = mapValue[newFeeAsset];
                vecSend[vecSend.size() - 1].nAmount = mapValue[lastRecipient.asset];

                if (!PreventDustForNewOutput(mapValue, vecSend, newFeeRecipient)) {
                    strFailReason = _("Output with the same asset as the fee could not find another asset which could be deducted from to make it not dust.");
                    return std::vector<CWalletTx>();
                }
            }
```

**Comment by AstryX** [comment]
General description of the code:
* Added sendanytoaddress rpc which combines multiple non-policy assets to satisfy a sent sum. It implements sendanymoney method which selects balances to be used, constructs and commits the transaction.
* New logic implemented in choosing an asset if feeAsset is undefined in CreateTransaction. It choses one with the largest balance and then subtracts from its outputs to fill the last change balance.
* Lots of edge case logic implemented if SelectCoins fails where it tries to find a new balance to supplement the missing amount (this may happen if CreateTransaction has enough for outputs at first but the fee pushes it over the limit). It subtracts as much as it can from the fee balance to utilize the newly found balance. This is done in order to give as much of the largest balance as possible.
* Balances chosen for sendany are sorted either ascending or descending. Descending is the default (largest first).
* There's recursive logic to split large transactions where it splits inputs between two different sets, constructs outputs and passes a set of remaining outputs to be used if the new transactions need them to pay for the fees.

**Comment by AstryX** [comment]
> Ok @AstryX makes sense. One more question;
> 
> Why do you exchange balance between last asset and fee asset:
> 
> ```
>             CRecipient lastRecipient = vecSend.back();
>             if (lastRecipient.asset != newFeeAsset) {
>                 CAmount lastBalance = balanceMap[lastRecipient.asset];
>                 CAmount usedLastBalance = mapValue[lastRecipient.asset];
>                 CAmount unusedLastBalance = lastBalance - usedLastBalance;
> 
>                 mapValue[lastRecipient.asset] = lastBalance;
>                 mapValue[newFeeAsset] -= unusedLastBalance;
> 
>                 vecSend[newFeeRecipient].nAmount = mapValue[newFeeAsset];
>                 vecSend[vecSend.size() - 1].nAmount = mapValue[lastRecipient.asset];
> 
>                 if (!PreventDustForNewOutput(mapValue, vecSend, newFeeRecipient)) {
>                     strFailReason = _("Output with the same asset as the fee could not find another asset which could be deducted from to make it not dust.");
>                     return std::vector<CWalletTx>();
>                 }
>             }
> ```

This is done in several places for three reasons. One is to make sure that the balance which is used for the fee is freed up as much as possible, thus can support the largest fee. The second reason is to minimize the amount of change that we are going to create. And the final reason is that we assume that it is very likely that the found feeAsset is going to be fully used up, thus we need to free some of it initially to even pay for the fee to begin with.

**Comment by nkostoulas** [comment]
> > Ok @AstryX makes sense. One more question;
> > Why do you exchange balance between last asset and fee asset:
> > ```
> >             CRecipient lastRecipient = vecSend.back();
> >             if (lastRecipient.asset != newFeeAsset) {
> >                 CAmount lastBalance = balanceMap[lastRecipient.asset];
> >                 CAmount usedLastBalance = mapValue[lastRecipient.asset];
> >                 CAmount unusedLastBalance = lastBalance - usedLastBalance;
> > 
> >                 mapValue[lastRecipient.asset] = lastBalance;
> >                 mapValue[newFeeAsset] -= unusedLastBalance;
> > 
> >                 vecSend[newFeeRecipient].nAmount = mapValue[newFeeAsset];
> >                 vecSend[vecSend.size() - 1].nAmount = mapValue[lastRecipient.asset];
> > 
> >                 if (!PreventDustForNewOutput(mapValue, vecSend, newFeeRecipient)) {
> >                     strFailReason = _("Output with the same asset as the fee could not find another asset which could be deducted from to make it not dust.");
> >                     return std::vector<CWalletTx>();
> >                 }
> >             }
> > ```
> 
> This is done in several places for three reasons. One is to make sure that the balance which is used for the fee is freed up as much as possible, thus can support the largest fee. The second reason is to minimize the number of changes that we are going to create. And the final reason is that we assume that it is very likely that the found feeAsset is going to be fully used up, thus we need to free some of it initially to even pay for the fee to begin with.

But this seems to be doing the opposite though. When the last asset is not the fee asset it means that the fee asset is the largest balance. Why would we remove from the largest balance and add to the smallest balance?

**Comment by AstryX** [comment]
> > > Ok @AstryX makes sense. One more question;
> > > Why do you exchange balance between last asset and fee asset:
> > > ```
> > >             CRecipient lastRecipient = vecSend.back();
> > >             if (lastRecipient.asset != newFeeAsset) {
> > >                 CAmount lastBalance = balanceMap[lastRecipient.asset];
> > >                 CAmount usedLastBalance = mapValue[lastRecipient.asset];
> > >                 CAmount unusedLastBalance = lastBalance - usedLastBalance;
> > > 
> > >                 mapValue[lastRecipient.asset] = lastBalance;
> > >                 mapValue[newFeeAsset] -= unusedLastBalance;
> > > 
> > >                 vecSend[newFeeRecipient].nAmount = mapValue[newFeeAsset];
> > >                 vecSend[vecSend.size() - 1].nAmount = mapValue[lastRecipient.asset];
> > > 
> > >                 if (!PreventDustForNewOutput(mapValue, vecSend, newFeeRecipient)) {
> > >                     strFailReason = _("Output with the same asset as the fee could not find another asset which could be deducted from to make it not dust.");
> > >                     return std::vector<CWalletTx>();
> > >                 }
> > >             }
> > > ```
> > 
> > 
> > This is done in several places for three reasons. One is to make sure that the balance which is used for the fee is freed up as much as possible, thus can support the largest fee. The second reason is to minimize the number of changes that we are going to create. And the final reason is that we assume that it is very likely that the found feeAsset is going to be fully used up, thus we need to free some of it initially to even pay for the fee to begin with.
> 
> But this seems to be doing the opposite though. When the last asset is not the fee asset it means that the fee asset is the largest balance. Why would we remove from the largest balance and add to the smallest balance?

Because, as I stated previously, the largest balance would likely have an output which fully spends it. This would throw an error as the fee needs some of it as well. So instead of nitpicking and removing a bit of the output at a time for every time fee changes we just free up as much of it as we can. By transfering it to the last one we avoid having two possible changes. It does not do the opposite.

**Comment by nkostoulas** [comment]
> > > > Ok @AstryX makes sense. One more question;
> > > > Why do you exchange balance between last asset and fee asset:
> > > > ```
> > > >             CRecipient lastRecipient = vecSend.back();
> > > >             if (lastRecipient.asset != newFeeAsset) {
> > > >                 CAmount lastBalance = balanceMap[lastRecipient.asset];
> > > >                 CAmount usedLastBalance = mapValue[lastRecipient.asset];
> > > >                 CAmount unusedLastBalance = lastBalance - usedLastBalance;
> > > > 
> > > >                 mapValue[lastRecipient.asset] = lastBalance;
> > > >                 mapValue[newFeeAsset] -= unusedLastBalance;
> > > > 
> > > >                 vecSend[newFeeRecipient].nAmount = mapValue[newFeeAsset];
> > > >                 vecSend[vecSend.size() - 1].nAmount = mapValue[lastRecipient.asset];
> > > > 
> > > >                 if (!PreventDustForNewOutput(mapValue, vecSend, newFeeRecipient)) {
> > > >                     strFailReason = _("Output with the same asset as the fee could not find another asset which could be deducted from to make it not dust.");
> > > >                     return std::vector<CWalletTx>();
> > > >                 }
> > > >             }
> > > > ```
> > > 
> > > 
> > > This is done in several places for three reasons. One is to make sure that the balance which is used for the fee is freed up as much as possible, thus can support the largest fee. The second reason is to minimize the number of changes that we are going to create. And the final reason is that we assume that it is very likely that the found feeAsset is going to be fully used up, thus we need to free some of it initially to even pay for the fee to begin with.
> > 
> > 
> > But this seems to be doing the opposite though. When the last asset is not the fee asset it means that the fee asset is the largest balance. Why would we remove from the largest balance and add to the smallest balance?
> 
> Because, as I stated previously, the largest balance would likely have an output which fully spends it. This would throw an error as the fee needs some of it as well. So instead of nitpicking and removing a bit of the output at a time for every time fee changes we just free up as much of it as we can. By transfering it to the last one we avoid having two possible changes. It does not do the opposite.

Alright ok. This all sounds a bit theoretical as I don't think the test covers even half of it. Any suggestions for what I **must** test?

**Comment by AstryX** [comment]
> > > > > Ok @AstryX makes sense. One more question;
> > > > > Why do you exchange balance between last asset and fee asset:
> > > > > ```
> > > > >             CRecipient lastRecipient = vecSend.back();
> > > > >             if (lastRecipient.asset != newFeeAsset) {
> > > > >                 CAmount lastBalance = balanceMap[lastRecipient.asset];
> > > > >                 CAmount usedLastBalance = mapValue[lastRecipient.asset];
> > > > >                 CAmount unusedLastBalance = lastBalance - usedLastBalance;
> > > > > 
> > > > >                 mapValue[lastRecipient.asset] = lastBalance;
> > > > >                 mapValue[newFeeAsset] -= unusedLastBalance;
> > > > > 
> > > > >                 vecSend[newFeeRecipient].nAmount = mapValue[newFeeAsset];
> > > > >                 vecSend[vecSend.size() - 1].nAmount = mapValue[lastRecipient.asset];
> > > > > 
> > > > >                 if (!PreventDustForNewOutput(mapValue, vecSend, newFeeRecipient)) {
> > > > >                     strFailReason = _("Output with the same asset as the fee could not find another asset which could be deducted from to make it not dust.");
> > > > >                     return std::vector<CWalletTx>();
> > > > >                 }
> > > > >             }
> > > > > ```
> > > > 
> > > > 
> > > > This is done in several places for three reasons. One is to make sure that the balance which is used for the fee is freed up as much as possible, thus can support the largest fee. The second reason is to minimize the number of changes that we are going to create. And the final reason is that we assume that it is very likely that the found feeAsset is going to be fully used up, thus we need to free some of it initially to even pay for the fee to begin with.
> > > 
> > > 
> > > But this seems to be doing the opposite though. When the last asset is not the fee asset it means that the fee asset is the largest balance. Why would we remove from the largest balance and add to the smallest balance?
> > 
> > 
> > Because, as I stated previously, the largest balance would likely have an output which fully spends it. This would throw an error as the fee needs some of it as well. So instead of nitpicking and removing a bit of the output at a time for every time fee changes we just free up as much of it as we can. By transfering it to the last one we avoid having two possible changes. It does not do the opposite.
> 
> Alright ok. This all sounds a bit theoretical as I don't think the test covers even half of it. Any suggestions for what I **must** test?

It does test a lot, in fact. I am not sure what else could be tested. It checks how it handles dust assets, sendany where fee makes inputs insufficient, different ways of sorting balances. What does not a lot more testing is splittransactions.

**Comment by nkostoulas** [comment]
> > > > > > Ok @AstryX makes sense. One more question;
> > > > > > Why do you exchange balance between last asset and fee asset:
> > > > > > ```
> > > > > >             CRecipient lastRecipient = vecSend.back();
> > > > > >             if (lastRecipient.asset != newFeeAsset) {
> > > > > >                 CAmount lastBalance = balanceMap[lastRecipient.asset];
> > > > > >                 CAmount usedLastBalance = mapValue[lastRecipient.asset];
> > > > > >                 CAmount unusedLastBalance = lastBalance - usedLastBalance;
> > > > > > 
> > > > > >                 mapValue[lastRecipient.asset] = lastBalance;
> > > > > >                 mapValue[newFeeAsset] -= unusedLastBalance;
> > > > > > 
> > > > > >                 vecSend[newFeeRecipient].nAmount = mapValue[newFeeAsset];
> > > > > >                 vecSend[vecSend.size() - 1].nAmount = mapValue[lastRecipient.asset];
> > > > > > 
> > > > > >                 if (!PreventDustForNewOutput(mapValue, vecSend, newFeeRecipient)) {
> > > > > >                     strFailReason = _("Output with the same asset as the fee could not find another asset which could be deducted from to make it not dust.");
> > > > > >                     return std::vector<CWalletTx>();
> > > > > >                 }
> > > > > >             }
> > > > > > ```
> > > > > 
> > > > > 
> > > > > This is done in several places for three reasons. One is to make sure that the balance which is used for the fee is freed up as much as possible, thus can support the largest fee. The second reason is to minimize the number of changes that we are going to create. And the final reason is that we assume that it is very likely that the found feeAsset is going to be fully used up, thus we need to free some of it initially to even pay for the fee to begin with.
> > > > 
> > > > 
> > > > But this seems to be doing the opposite though. When the last asset is not the fee asset it means that the fee asset is the largest balance. Why would we remove from the largest balance and add to the smallest balance?
> > > 
> > > 
> > > Because, as I stated previously, the largest balance would likely have an output which fully spends it. This would throw an error as the fee needs some of it as well. So instead of nitpicking and removing a bit of the output at a time for every time fee changes we just free up as much of it as we can. By transfering it to the last one we avoid having two possible changes. It does not do the opposite.
> > 
> > 
> > Alright ok. This all sounds a bit theoretical as I don't think the test covers even half of it. Any suggestions for what I **must** test?
> 
> It does test a lot, in fact. It might be not obvious at first glance. I am not sure what else could be tested. It checks how it handles dust assets, sendany where fee makes inputs insufficient, different ways of sorting balances. What does not a lot more testing is splittransactions.

Ok thanks for the input!

**Comment by AstryX** [comment]
> > > > > > Ok @AstryX makes sense. One more question;
> > > > > > Why do you exchange balance between last asset and fee asset:
> > > > > > ```
> > > > > >             CRecipient lastRecipient = vecSend.back();
> > > > > >             if (lastRecipient.asset != newFeeAsset) {
> > > > > >                 CAmount lastBalance = balanceMap[lastRecipient.asset];
> > > > > >                 CAmount usedLastBalance = mapValue[lastRecipient.asset];
> > > > > >                 CAmount unusedLastBalance = lastBalance - usedLastBalance;
> > > > > > 
> > > > > >                 mapValue[lastRecipient.asset] = lastBalance;
> > > > > >                 mapValue[newFeeAsset] -= unusedLastBalance;
> > > > > > 
> > > > > >                 vecSend[newFeeRecipient].nAmount = mapValue[newFeeAsset];
> > > > > >                 vecSend[vecSend.size() - 1].nAmount = mapValue[lastRecipient.asset];
> > > > > > 
> > > > > >                 if (!PreventDustForNewOutput(mapValue, vecSend, newFeeRecipient)) {
> > > > > >                     strFailReason = _("Output with the same asset as the fee could not find another asset which could be deducted from to make it not dust.");
> > > > > >                     return std::vector<CWalletTx>();
> > > > > >                 }
> > > > > >             }
> > > > > > ```
> > > > > 
> > > > > 
> > > > > This is done in several places for three reasons. One is to make sure that the balance which is used for the fee is freed up as much as possible, thus can support the largest fee. The second reason is to minimize the number of changes that we are going to create. And the final reason is that we assume that it is very likely that the found feeAsset is going to be fully used up, thus we need to free some of it initially to even pay for the fee to begin with.
> > > > 
> > > > 
> > > > But this seems to be doing the opposite though. When the last asset is not the fee asset it means that the fee asset is the largest balance. Why would we remove from the largest balance and add to the smallest balance?
> > > 
> > > 
> > > Because, as I stated previously, the largest balance would likely have an output which fully spends it. This would throw an error as the fee needs some of it as well. So instead of nitpicking and removing a bit of the output at a time for every time fee changes we just free up as much of it as we can. By transfering it to the last one we avoid having two possible changes. It does not do the opposite.
> > 
> > 
> > Alright ok. This all sounds a bit theoretical as I don't think the test covers even half of it. Any suggestions for what I **must** test?
> 
> It does test a lot, in fact. It might be not obvious at first glance. I am not sure what else could be tested. It checks how it handles dust assets, sendany where fee makes inputs insufficient, different ways of sorting balances. What does not a lot more testing is splittransactions.

One potential edge case which I checked that it worked, but would need a reset of the balances is importing two assets 5.0 each and then spending 5. The coinselection would fail and it would look for a new asset. It would find another 5.0 and transfer all of the first one to the second one, however, it would leave a 0 output on the first one, thus, causing an error. This is resolved in code, and I have checked to make sure it works, but it does not have a permanent test.

**Comment by AstryX** [comment]
No worries!

**Comment by nkostoulas** [comment]
Any clues as to why the split tx test might be failing?

**Comment by nkostoulas** [comment]
> > > > > > > Ok @AstryX makes sense. One more question;
> > > > > > > Why do you exchange balance between last asset and fee asset:
> > > > > > > ```
> > > > > > >             CRecipient lastRecipient = vecSend.back();
> > > > > > >             if (lastRecipient.asset != newFeeAsset) {
> > > > > > >                 CAmount lastBalance = balanceMap[lastRecipient.asset];
> > > > > > >                 CAmount usedLastBalance = mapValue[lastRecipient.asset];
> > > > > > >                 CAmount unusedLastBalance = lastBalance - usedLastBalance;
> > > > > > > 
> > > > > > >                 mapValue[lastRecipient.asset] = lastBalance;
> > > > > > >                 mapValue[newFeeAsset] -= unusedLastBalance;
> > > > > > > 
> > > > > > >                 vecSend[newFeeRecipient].nAmount = mapValue[newFeeAsset];
> > > > > > >                 vecSend[vecSend.size() - 1].nAmount = mapValue[lastRecipient.asset];
> > > > > > > 
> > > > > > >                 if (!PreventDustForNewOutput(mapValue, vecSend, newFeeRecipient)) {
> > > > > > >                     strFailReason = _("Output with the same asset as the fee could not find another asset which could be deducted from to make it not dust.");
> > > > > > >                     return std::vector<CWalletTx>();
> > > > > > >                 }
> > > > > > >             }
> > > > > > > ```
> > > > > > 
> > > > > > 
> > > > > > This is done in several places for three reasons. One is to make sure that the balance which is used for the fee is freed up as much as possible, thus can support the largest fee. The second reason is to minimize the number of changes that we are going to create. And the final reason is that we assume that it is very likely that the found feeAsset is going to be fully used up, thus we need to free some of it initially to even pay for the fee to begin with.
> > > > > 
> > > > > 
> > > > > But this seems to be doing the opposite though. When the last asset is not the fee asset it means that the fee asset is the largest balance. Why would we remove from the largest balance and add to the smallest balance?
> > > > 
> > > > 
> > > > Because, as I stated previously, the largest balance would likely have an output which fully spends it. This would throw an error as the fee needs some of it as well. So instead of nitpicking and removing a bit of the output at a time for every time fee changes we just free up as much of it as we can. By transfering it to the last one we avoid having two possible changes. It does not do the opposite.
> > > 
> > > 
> > > Alright ok. This all sounds a bit theoretical as I don't think the test covers even half of it. Any suggestions for what I **must** test?
> > 
> > 
> > It does test a lot, in fact. It might be not obvious at first glance. I am not sure what else could be tested. It checks how it handles dust assets, sendany where fee makes inputs insufficient, different ways of sorting balances. What does not a lot more testing is splittransactions.
> 
> One potential edge case which I checked that it worked, but would need a reset of the balances is importing two assets 5.0 each and then spending 5. The coinselection would fail and it would look for a new asset. It would find another 5.0 and transfer all of the first one to the second one, however, it would leave a 0 output on the first one, thus, causing an error. This is resolved in code, and I have checked to make sure it works, but it does not have a permanent test.

Any clues as to why the split tx test might be failing? Or where in the code I could look at.

File: src/wallet/wallet.cpp [hunk 1]
```diff
@@ -3402,8 +3560,201 @@ bool CWallet::CreateTransaction(const vector<CRecipient>& vecSend, CWalletTx& wt
...
+                for (const auto& recipient : vecSend)
+                {
+                    if(recipientMap.find(recipient.asset) != recipientMap.end())
+                        recipientMap[recipient.asset].push_back(recipient);
+                    else {
+                        std::vector<CRecipient> initVec;
+                        initVec.push_back(recipient);
+                        recipientMap[recipient.asset] = initVec;
+                    }
+                    mapValue[recipient.asset] += recipient.nAmount;
```
Comment (nkostoulas) [inline src/wallet/wallet.cpp]: why this?

**Comment by nkostoulas** [comment]
Maybe it's failing because the code references `mAvailableInputs` which is undefined.

**Comment by nkostoulas** [comment]
I take that back. `mAvailableInputs` is indeed set but the outputs already used are removed, which makes no sense as when splitting we might want to re-use them no?

File: src/wallet/wallet.cpp (hunk 1 above)
Comment (AstryX) [reply src/wallet/wallet.cpp]: Probably a mistake. the "mapValue[recipient.asset] += recipient.nAmount;" does not seem to do anything productive.

**Comment by AstryX** [comment]
> I take that back. `mAvailableInputs` is indeed set but the outputs already used are removed, which makes no sense as when splitting we might want to re-use them no?

No, that is wrong. It does make sense to remove them as the logic makes sure that the first split transaction only gets enough inputs to satisfy the deficit. The second transaction is then free to choose from the remaining pool of inputs. It is true that we might want to use a single input in multiple locations, however, these transactions are created simultaneously thus we can't put the same input in two transactions as it will change after the first tx gets committed (change). Naturally, there are edge cases to consider for this logic too, but, frankly, I prioritized my time for sendany so this is not as bulletproof.

**Comment by nkostoulas** [comment]
The code was failing because of this `auto &freeInputVector = mAvailableInputs->at(newAsset)`. I added checks that the asset is there and now it no longer fails.

This means that something is missing for mAvailableInputs. Shouldn't mAvailableInputs be all inputs? What do you mean the first gets enough and the second is free? Recipients are split, inputs are split and mAvailableInputs is the same for both left and right.

File: src/wallet/wallet.cpp [hunk 2]
```diff
@@ -3402,8 +3560,201 @@ bool CWallet::CreateTransaction(const vector<CRecipient>& vecSend, CWalletTx& wt
...
+                for (unsigned int i = 0; i < vChangeKeys[0].size(); ++i) {
+                    vChangeKeys[0][i].ReturnKey();
+                }
+                // Create and send the transaction
+                std::vector<std::vector<CReserveKey>> vChangeKeysLeft;
+                std::vector<std::vector<CReserveKey>> vChangeKeysRight;
+                std::vector<CReserveKey> vChangeLeft;
+                std::vector<CReserveKey> vChangeRight;
+                vChangeLeft.reserve(leftRecipients.size() + 1);
+                for (unsigned int j = 0; j < vecSend.size() + 1; ++j) {
```
Comment (nkostoulas) [inline src/wallet/wallet.cpp]: shouldn't this be `leftRecipients.size()`?

File: src/wallet/wallet.cpp [hunk 3]
```diff
@@ -3402,8 +3560,201 @@ bool CWallet::CreateTransaction(const vector<CRecipient>& vecSend, CWalletTx& wt
...
+                    vChangeLeft.emplace_back(pwalletMain);
+                }
+                vChangeKeysLeft.push_back(vChangeLeft);
+
+                vChangeRight.reserve(rightRecipients.size() + 1);
+                for (unsigned int j = 0; j < vecSend.size() + 1; ++j) {
```
Comment (nkostoulas) [inline src/wallet/wallet.cpp]: and this `rightRecipients.size()`?

**Comment by AstryX** [comment]
> The code was failing because of this `auto &freeInputVector = mAvailableInputs->at(newAsset)`. The asset was missing from mAvailableInputs. I added checks that the asset is there and now it no longer fails.
> 
> The error now is
> `JSONRPC error: Selecting coins error. Amount needed was not larger than the amount owned.`
> 
> This means that something is missing for mAvailableInputs. Shouldn't mAvailableInputs be all inputs? What do you mean the first gets enough and the second is free? Recipients are split, inputs are split and mAvailableInputs is the same for both left and right.
> 
> ```
> std::vector<CWalletTx> leftRes = CreateTransaction(leftRecipients, dummyLeft, vChangeKeysLeft,
>                     feeLeft, dummyPosLeft, strFailReason, coinControl, true, NULL, true, NULL, NULL, NULL,
>                     CAsset(), fIgnoreBlindFail, fSplitTransactions, leftInputs, fFindFeeAsset, mAvailableInputs);
> 
> std::vector<CWalletTx> rightRes = CreateTransaction(rightRecipients, dummyRight, vChangeKeysRight,
>                     feeRight, dummyPosRight, strFailReason, coinControl, true, NULL, true, NULL, NULL, NULL,
>                     CAsset(), fIgnoreBlindFail, fSplitTransactions, rightInputs, fFindFeeAsset, mAvailableInputs);
> ```

mAvailableInputs stores inputs that the wallet still has which are not part of the initial transaction which was deemed too large. It is done this way so transactions can't choose the same inputs if they search for additional ones. It should not have all of the inputs. The error is odd, it just says that the amount needed is not > amount have. This error should never be hit unless something is broken along the way.

File: src/wallet/wallet.cpp (hunk 3 above)
Comment (AstryX) [reply src/wallet/wallet.cpp]: Yes for both, that is one of the last bits of code I wrote before leaving, must have made a mistake which I could not fully test due to the tests failing.

**Comment by nkostoulas** [comment]
Ok so it seems that coin select failed even though we have enough coins.

**Comment by nkostoulas** [comment]
if mAvailableInputs has only the unused ones though, wouldn't the code below swap the top asset with a balance not in your coins used?

```
    if (feeAsset.IsNull())
    {
        CAsset newFeeAsset = vecSend[0].asset;
        if (vecSend.size() > 1) {
            CAmountMap balanceMap;
            if (mAvailableInputs && mAvailableInputs->size() > 0) {
                balanceMap = GetBalanceFromOutputs(*mAvailableInputs);
            } else {
                balanceMap = pwalletMain->GetBalance();
            }
            int newFeeRecipient = 0;
            for (unsigned int i = 1; i < vecSend.size(); ++i) {
                if (balanceMap[vecSend[i].asset] > balanceMap[newFeeAsset]) {
                    newFeeAsset = vecSend[i].asset;
                    newFeeRecipient = i;
                }
            }

            // I don't understand the purpose of this
            // maybe in case a recipient is dust, which only the last can be?
            // maybe SendAnyMoney should handle that
            CRecipient lastRecipient = vecSend.back();
            if (lastRecipient.asset != newFeeAsset) {
                CAmount lastBalance = balanceMap[lastRecipient.asset];
                CAmount usedLastBalance = mapValue[lastRecipient.asset];
                CAmount unusedLastBalance = lastBalance - usedLastBalance;

                mapValue[lastRecipient.asset] = lastBalance;
                mapValue[newFeeAsset] -= unusedLastBalance;

                vecSend[newFeeRecipient].nAmount = mapValue[newFeeAsset];
                vecSend[vecSend.size() - 1].nAmount = mapValue[lastRecipient.asset];

                if (!PreventDustForNewOutput(mapValue, vecSend, newFeeRecipient)) {
                    strFailReason = _("Output with the same asset as the fee could not find another asset which could be deducted from to make it not dust.");
                    return std::vector<CWalletTx>();
                }
            }
        }
        feeAsset = newFeeAsset;
    }
```

**Comment by AstryX** [comment]
> balanceMap = GetBalanceFromOutputs(*mAvailableInputs);

Yes, that is a big error, that function should use vInputPool which contains the inputs that the split transactions should use for their outputs. Thanks for finding it!!

**Comment by AstryX** [comment]
vInputPool is created by splitting inputs of the previous transaction which was too large. It is divided into left and right to make sure that the two new txs do not use overlapping inputs. Then the split transactions are created and tested within CreateTransaction function. If it fails due to insufficient balance at that point, then mAvailableInputs is used to make each transaction valid. Map is shared, so if an input is used by one of the split tx's, then it is removed from the map.

**Comment by nkostoulas** [comment]
Also I think either left or right recipients will contain a recipient with the fee subtracted. This will create a mismatch between recipient and input (from vInputPool) causing the error I sent. I think I need to re-add the fee before splitting.

**Comment by nkostoulas** [comment]
Or maybe it's not the fee but the result of DistributeFeeToNewBalance that needs to be undone.

**Comment by nkostoulas** [comment]
I'll try keep the original vecSend and use that when splitting.

**Comment by nkostoulas** [comment]
Alright, I had to use the original unchanged `vecSend` when splitting, because before the `vecSend` vector was modified in an attempt to pay the tx fee and that created inconsistencies between `inputPool` and `recipients` and splitting. 

I am not sure if this will have any implications where a new input of the same address has been introduced but that should not happen because if `vecSend` changes it's only for a new asset. Also I don't understand why we are creating a map of asset to vector of recipients. Isn't it only one asset one recipient?

**Comment by AstryX** [comment]
> Alright, I had to use the original unchanged `vecSend` when splitting, because before the `vecSend` vector was modified in an attempt to pay the tx fee and that created inconsistencies between `inputPool` and `recipients` and splitting.
> 
> I am not sure if this will have any implications where a new input of the same address has been introduced but that should not happen because if `vecSend` changes it's only for a new asset. Also I don't understand why we are creating a map of asset to vector of recipients. Isn't it only one asset one recipient?

Alright man, do what you think is best :) The reason for that is that the code was initially written to support send**many**. It does not seem like splitting will be supported for that though.

**Comment by nkostoulas** [comment]
Well it didn’t work before. Now using thenoriginal vecSend sendany works with split tx.

Still I think there is an edge case where these inconsistencies between vecSend and inputPool can still occur and code might require some redesign.

**Comment by nkostoulas** [comment]
Oh i didn’t know about sendmany. Which part of the code? Maybe I should remove it.

**Comment by nkostoulas** [comment]
By the way currently choosing as feeAsset the asset with the largest balance, means that all the coins of this asset are already chosen so SelectCoins() will always fail and a new asset will need to be chosen for fees. I don't see the point of this.

**Pull request merged**